### PR TITLE
[MRG] ENH: Update `laplacian` function introducing the new `copy` option

### LINKED ISSUE #10826: Graph Laplacian sparse format change ruins the in-place calculation
**Is your feature request related to a problem? Please describe.**
Since https://github.com/scipy/scipy/pull/15251 the function https://docs.scipy.org/doc/scipy-1.3.0/reference/generated/scipy.sparse.csgraph.laplacian.html now support in-place calculation, without doubling the amount of memory on the output , cf. https://github.com/scikit-learn/scikit-learn/pull/14636

However, it appears that the code is not preserving the format other than COO of the graph sparse matrix - the implementation explicitly makes a copy unless the input format is COO, as at least now documented in https://github.com/scipy/scipy/pull/15251.  It's really unclear why. 

**Describe the solution you'd like**
In-place and strictly format- and type-preserving calculation with `copy=False` avoiding the memory usage spikes. Cf. https://github.com/scipy/scipy/issues/10827

### PR DESCRIPTION
**Pull request by lobpcg**
introduce the new copy option, to save 1/2 memory with `copy=False`

#### Reference issue
Partially closes #10826 and https://github.com/scikit-learn/scikit-learn/issues/8834

#### What does this implement/fix?
 introduce the new copy option

### CODE REVIEW

File: scipy/sparse/csgraph/_laplacian.py
Comment (tupui) [inline scipy/sparse/csgraph/_laplacian.py]:
```suggestion
    use_out_degree=False,
    *,
    inplace=False,
):
```

**Review by tupui — COMMENTED** [review]
Thanks @lobpcg. Good start, I think we should test the parameter. And looking at the discussion in the issue, it would be good to document the copying mechanism in more details.

File: scipy/sparse/csgraph/_laplacian.py
Comment (lobpcg) [reply scipy/sparse/csgraph/_laplacian.py]: Thanks, have not seen your suggestion and fixed differently

Comment (tupui) [reply scipy/sparse/csgraph/_laplacian.py]: No problem, just need to add the keyword only identifier.

File: scipy/sparse/csgraph/_laplacian.py
Comment (tupui) [inline scipy/sparse/csgraph/_laplacian.py:17]:
```suggestion
def laplacian(csgraph, normed=False, return_diag=False, use_out_degree=False,
              *, inplace=False):
```

Comment (lobpcg) [reply scipy/sparse/csgraph/_laplacian.py:17]: why? I do not understand

**Comment by lobpcg** [comment]
> Thanks @lobpcg. Good start, I think we should test the parameter. And looking at the discussion in the issue, it would be good to document the copying mechanism in more details.

If you could suggest a unit test, I would appreciate it. I'll update the documentation.

**Comment by tupui** [comment]
> If you could suggest a unit test, I would appreciate it. I'll update the documentation.

I am on the road with only a phone for another week. I would do something simple like passing an array and check that the original array is either modified or not.

**Comment by lobpcg** [comment]
> > If you could suggest a unit test, I would appreciate it. I'll update the documentation.
> 
> I am on the road with only a phone for another week. I would do something simple like passing a graph and check that the original graph is either modified or not.

Yes, I thought about it and it's easy to code, but this does not explicitly checks the memory usage.

**Comment by lobpcg** [comment]
Could anyone please advice? I do not know how to properly format to pass values in 
```
@pytest.mark.parametrize("dtype", [np.float32, np.float64])
@pytest.mark.parametrize("inplace", ["FASLE", "TRUE"])
@pytest.mark.parametrize("normed", ["FASLE", "TRUE"])
@pytest.mark.parametrize("use_out_degree", ["FASLE", "TRUE"])
```
so getting errors:
```
scipy/sparse/csgraph/tests/test_graph_laplacian.py:107:18: F821 undefined name 'dtype'
scipy/sparse/csgraph/tests/test_graph_laplacian.py:109:12: F821 undefined name 'normed'
scipy/sparse/csgraph/tests/test_graph_laplacian.py:109:23: F821 undefined name 'use_out_degree'
scipy/sparse/csgraph/tests/test_graph_laplacian.py:116:8: F821 undefined name 'normed'
scipy/sparse/csgraph/tests/test_graph_laplacian.py:116:19: F821 undefined name 'use_out_degree'
scipy/sparse/csgraph/tests/test_graph_laplacian.py:123:12: F821 undefined name 'normed'
scipy/sparse/csgraph/tests/test_graph_laplacian.py:123:27: F821 undefined name 'use_out_degree'
scipy/sparse/csgraph/tests/test_graph_laplacian.py:130:8: F821 undefined name 'normed'
scipy/sparse/csgraph/tests/test_graph_laplacian.py:130:23: F821 undefined name 'use_out_degree'
scipy/sparse/csgraph/tests/test_graph_laplacian.py:137:18: F821 undefined name 'dtype'
scipy/sparse/csgraph/tests/test_graph_laplacian.py:138:18: F821 undefined name 'dtype'
scipy/sparse/csgraph/tests/test_graph_laplacian.py:140:29: F821 undefined name 'normed'
scipy/sparse/csgraph/tests/test_graph_laplacian.py:141:37: F821 undefined name 'use_out_degree'
scipy/sparse/csgraph/tests/test_graph_laplacian.py:142:30: F821 undefined name 'inplace'
```

**Comment by lobpcg** [comment]
The unit tests unfortunately require some deep and time consuming reorganization, replacing loops with @pytest.mark.parametrize, since testing inplace cannot evidently be run in a loop...

**Comment by lobpcg** [comment]
There are so many clearly unrelated to this PR errors that I am unsure where to find or to check the lack of the relevant errors...

**Comment by tupui** [comment]
There are some issues with other dependencies right now which should be fixed very soon.

File: scipy/sparse/csgraph/_laplacian.py
Comment (tupui) [reply scipy/sparse/csgraph/_laplacian.py:17]: This is to make it keyword only. When we add new parameters, we do this now so we are free to move them around. It's also clearer in user code.

File: scipy/sparse/csgraph/_laplacian.py
Comment (tupui) [inline scipy/sparse/csgraph/_laplacian.py:142]: Why not just
```suggestion
    m = np.asarray(graph)
```

File: scipy/sparse/csgraph/tests/test_graph_laplacian.py
Comment (tupui) [inline scipy/sparse/csgraph/tests/test_graph_laplacian.py:13]: As you did bellow, this could also benefit from some parametrize.

File: scipy/sparse/csgraph/_laplacian.py
Comment (lobpcg) [reply scipy/sparse/csgraph/_laplacian.py:142]: This is the only place where `inplace` is used. If deleted, `inplace` would just have no effect?

File: scipy/sparse/csgraph/tests/test_graph_laplacian.py
Comment (lobpcg) [reply scipy/sparse/csgraph/tests/test_graph_laplacian.py:13]: I do not actually understand what this does, so have not changed it, just moved for better code readability. The goal of this PR is only to add `inplace` and some tests for this, not to beautify/re-write all unit tests.

File: scipy/sparse/csgraph/_laplacian.py
Comment (tupui) [inline scipy/sparse/csgraph/_laplacian.py:56]: This needs double backtick because it's not just a reference to a parameter.
```suggestion
    Unless the ``inplace=True`` parameter is used, the constructed Laplacian
```

File: scipy/sparse/csgraph/_laplacian.py
Comment (tupui) [inline scipy/sparse/csgraph/_laplacian.py:59]: I would talk about type. Or use double backticks.
```suggestion
    The Laplacian preserved the type and the format of the input
```

File: scipy/sparse/csgraph/_laplacian.py
Comment (tupui) [inline scipy/sparse/csgraph/_laplacian.py:61]:
```suggestion
    converted into `coo` thus doubling the memory even with ``inplace=True``
```
I am not sure about the rendering the matrix types. It should at least be consistent and if possible link.

File: scipy/sparse/csgraph/_laplacian.py
Comment (tupui) [inline scipy/sparse/csgraph/_laplacian.py:69]:
```suggestion
    replaced with zeros for the purpose of normalization where ``normed=True``.
```

File: scipy/sparse/csgraph/_laplacian.py
Comment (tupui) [reply scipy/sparse/csgraph/_laplacian.py:144]: Would need some tests, but `asarray` should not do anything if it's already an array (the case here). To me it should be equivalent.

File: scipy/sparse/csgraph/tests/test_graph_laplacian.py
Comment (tupui) [reply scipy/sparse/csgraph/tests/test_graph_laplacian.py:13]: Ok you can leave this out. I am not a fan of this test as well because I don't know what `ValueError` this is referring to (does not have a match statement).

Comment (lobpcg) [reply scipy/sparse/csgraph/tests/test_graph_laplacian.py:13]: Maybe it checks for lacking the required input.

File: scipy/sparse/csgraph/_laplacian.py
Comment (lobpcg) [reply scipy/sparse/csgraph/_laplacian.py:144]: Your proposed edit would just get the original code back. Since the original is not inplace, I guess that m = np.array(graph) creates a copy if graph is already an array. I'll try to test it.

Comment (tupui) [reply scipy/sparse/csgraph/_laplacian.py:144]: My edit vs the original code is `asarray` instead of `array`

Comment (lobpcg) [reply scipy/sparse/csgraph/_laplacian.py]: Sorry, I missed `as`. You are correct. Testing it now.

Comment (lobpcg) [reply scipy/sparse/csgraph/_laplacian.py]: But I am also correct, since your edit removes any use of `inplace', so it has no effect.

Comment (lobpcg) [reply scipy/sparse/csgraph/_laplacian.py]: The correct version should be:
```
    if inplace:
            m = np.asarray(graph)
        else:
            m = np.array(graph)
```
Testing the above now

File: scipy/sparse/csgraph/_laplacian.py
Comment (lobpcg) [reply scipy/sparse/csgraph/_laplacian.py]: This is what the original code actually does. I have not changed it, only documented.
 
I have no clue why  sparse 'lil' and 'dok' formats are converted into `coo` but changing it should be a different PR also dealing with not being backward compatible.

File: scipy/sparse/csgraph/_laplacian.py
Comment (tupui) [reply scipy/sparse/csgraph/_laplacian.py]: Correct 👍 Just not sure about the `asarray` because it will not be inplace if there is a conversion to make (not a ndarray).

**Comment by lobpcg** [comment]
The dense part seems working - passing all tests. 

The space part `inline=True` is not actually inline so far. My current description of the conversion logic is inaccurate. It appears from the code that all sparse formats are converted to `coo`, except for `dia`.  The code logic is strange, so I am unsure if this is intentional or a bug.

File: scipy/sparse/csgraph/_laplacian.py
Comment (lobpcg) [reply scipy/sparse/csgraph/_laplacian.py]: My current description of the sparse format conversion logic is inaccurate. It appears from the code that all sparse formats are converted to coo, except for dia. The code logic is strange, so I am unsure if this is intentional or a bug.

**Comment by lobpcg** [comment]
I have renamed `inplace` to `copy` to simplify the logic and be consistent with other functions..

**Comment by lobpcg** [comment]
There is one test for dense matrices where `copy=False` has no effect, unsure why:
```
________________________ test_symmetric_graph_laplacian ________________________
scipy/sparse/csgraph/tests/test_graph_laplacian.py:79: in test_symmetric_graph_laplacian
    _check_symmetric_graph_laplacian(mat, normed, copy)
        copy       = False
        mat        = 'np.arange(10) * np.arange(10)[:, np.newaxis]'
        normed     = True
        symmetric_mats = ('np.arange(10) * np.arange(10)[:, np.newaxis]', 'np.ones((7, 7))', 'np.eye(19)', 'sparse.diags([1, 1], [-1, 1], shape... 'sparse.diags([1, 1], [-1, 1], shape=(4, 4)).toarray()', 'sparse.diags([1, 1], [-1, 1], shape=(4, 4)).todense()', ...)
scipy/sparse/csgraph/tests/test_graph_laplacian.py:62: in _check_symmetric_graph_laplacian
    assert_array_almost_equal(laplacian, mat)
E   AssertionError: 
E   Arrays are not almost equal to 6 decimals
E   
E   Mismatched elements: 80 / 100 (80%)
E   Max absolute difference: 80.
E   Max relative difference: 1.0162564
E    x: array([[ 0.      , -0.      , -0.      , -0.      , -0.      , -0.      ,
E           -0.      , -0.      , -0.      , -0.      ],
E          [-0.      ,  1.      , -0.032513, -0.040291, -0.047088, -0.0533  ,...
E    y: array([[ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
E          [ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
E          [ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18],...
        copy       = False
        explicit_laplacian = array([[ 0.        , -0.        , -0.        , -0.        , -0.        ,
        -0.        , -0.        , -0.        ...7784, -0.10783277, -0.13363062, -0.15617376,
        -0.1767767 , -0.19611614, -0.21459877, -0.23249528,  1.        ]])
        laplacian  = array([[ 0.        , -0.        , -0.        , -0.        , -0.        ,
        -0.        , -0.        , -0.        ...7784, -0.10783277, -0.13363062, -0.15617376,
        -0.1767767 , -0.19611614, -0.21459877, -0.23249528,  1.        ]])
        mat        = array([[ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
       [ 0,  2,  4, ...35, 42, 49, 56, 63],
       [ 0,  8, 16, 24, 32, 40, 48, 56, 64, 72],
       [ 0,  9, 18, 27, 36, 45, 54, 63, 72, 81]])
        n_nodes    = 10
        normed     = True
        sp_laplacian = <10x10 sparse matrix of type '<class 'numpy.float64'>'
	with 82 stored elements in COOrdinate format>
        sp_mat     = <10x10 sparse matrix of type '<class 'numpy.int64'>'
	with 81 stored elements in Compressed Sparse Row format>
        tested     = array([[ 0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
         0.        ,  0.        ,  0.        ...7784, -0.10783277, -0.13363062, -0.15617376,
        -0.1767767 , -0.19611614, -0.21459877, -0.23249528,  1.        ]])
```

**Comment by lobpcg** [comment]
It appears from the code that all sparse formats are converted to `coo`, except for `dia`. The code logic is strange, so I am unsure if this is intentional or a bug.

Anyway, PR  is basically completed and ready for review/merge.

**Comment by lobpcg** [comment]
There are more issues to address with this same script, which I could try doing after this PR is merged:

- https://github.com/scikit-learn/scikit-learn/issues/21274
- https://github.com/scipy/scipy/issues/10827
- https://github.com/scipy/scipy/issues/14490
- https://github.com/scipy/scipy/issues/10006
- https://github.com/scikit-learn/scikit-learn/issues/21274

**Comment by lobpcg** [comment]
LGTM, could someone please review/merge?

File: scipy/sparse/csgraph/tests/test_graph_laplacian.py
Comment (tupui) [inline scipy/sparse/csgraph/tests/test_graph_laplacian.py:61]: If I am not mistaking, there is no test checking that the original `mat` is unaltered when `copy=True`. You can do a copy as the first step to store the original value to compare with. As for the False case, the evidence that the array is different should happen before any other condition.

File: scipy/sparse/csgraph/_laplacian.py
Comment (tupui) [reply scipy/sparse/csgraph/_laplacian.py]: So what did you do in the end?

File: scipy/sparse/csgraph/tests/test_graph_laplacian.py
Comment (lobpcg) [reply scipy/sparse/csgraph/tests/test_graph_laplacian.py]: Correct. I have added a check for `copy=True` and swapped the order

Comment (lobpcg) [reply scipy/sparse/csgraph/tests/test_graph_laplacian.py]: @tupui Done. The failures appear unrelated to this PR.

File: scipy/sparse/csgraph/_laplacian.py
Comment (tupui) [inline scipy/sparse/csgraph/_laplacian.py:35]:
```suggestion
        If False, then change `csgraph` in place if possible,
```

File: scipy/sparse/csgraph/_laplacian.py
Comment (tupui) [inline scipy/sparse/csgraph/_laplacian.py:33]: Just seeing that we could add the defaults for `normed` and `return_diag`. Would be great if you could do this, I cannot do a suggestion as it's too far above.

File: scipy/sparse/csgraph/tests/test_graph_laplacian.py
Comment (tupui) [inline scipy/sparse/csgraph/tests/test_graph_laplacian.py:67]: If we don't copy, we should still test that `mat != mat_copy` to show that it was really inline. (we can either check that we would raise an `AssertionError` with `assert_allclose`)

File: scipy/sparse/csgraph/tests/test_graph_laplacian.py
Comment (tupui) [inline scipy/sparse/csgraph/tests/test_graph_laplacian.py:72]: Sorry just seeing this, but since we are touching these we should use `assert_allclose` instead as `assert_array_almost_equal` is not recommended anymore.

**Review by tupui — COMMENTED** [review]
Some very minor things and it should be good to go. Thanks for your patience in the process!

File: scipy/sparse/csgraph/tests/test_graph_laplacian.py
Comment (lobpcg) [reply scipy/sparse/csgraph/tests/test_graph_laplacian.py]: done

File: scipy/sparse/csgraph/tests/test_graph_laplacian.py
Comment (lobpcg) [reply scipy/sparse/csgraph/tests/test_graph_laplacian.py]: I believe such a check would be redundant, since I already test that `laplacian == mat` and `laplacian` is never the same as `mat_copy`, which always implies `mat != mat_copy`

File: scipy/sparse/csgraph/_laplacian.py
Comment (lobpcg) [reply scipy/sparse/csgraph/_laplacian.py:35]: done

**Comment by lobpcg** [comment]
@tupui - thanks for another review! 
Could you please resolve your conversations yourself where/after you feel that they were resolved?

**Review by tupui — APPROVED** [review]
Thanks again @lobpcg and sorry for the long process. I will wait for some CI to come back green (there has been a branch rename so circleCI is red for that, so not related).

**Comment by lobpcg** [comment]
`ci/circleci: build_scipy` appears broken at the moment. All other tests pass. LGTM

File: scipy/sparse/csgraph/_laplacian.py
Comment (lobpcg) [reply scipy/sparse/csgraph/_laplacian.py]: I just updated the docstring to reflect my understanding of what the code actually does with various sparse formats, without making any related changes in the code, so it's fully backward compatible.

**Pull request merged**
